contracts/contracts/Multisig.cairo

```diff
@@ -470,24 +432,111 @@ func execute_transaction{
     return (response_len=response.retdata_size, response=response.retdata)
 end
 
+# Sets number of required confirmations. The only way this can be invoked
+# is via a recursive call from execute_transaction -> set_confirmations_required.
+@external
+func set_confirmations_required{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    confirmations_required : felt
+):
+    require_multisig()
+    _set_confirmations_required_impl(confirmations_required)
+
+    return ()
+end
+
+# Sets the owners field on the multisig. The only way this can be invoked
+# is via a recursive call from execute_transaction -> set_owners.
+# Number of required confirmations is decreased in case its larger than
+# number of new owners.
+@external
+func set_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*
+):
+    require_multisig()
+    _set_owners_impl(owners_len, owners)
+
+    return ()
+end
+
+# Set new owners and number of required confirmations.
+# Can be called only via recursively from
+# execute_transaction -> set_owners_and_confirmations_required
+@external
+func set_owners_and_confirmations_required{
+    syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr
+}(owners_len : felt, owners : felt*, confirmations_required : felt):
+    set_owners(owners_len, owners)
+    set_confirmations_required(confirmations_required)
+
+    return ()
+end
+
 #
 # Storage Helpers
 #
 
-func _set_owners{
-        syscall_ptr : felt*,
-        pedersen_ptr : HashBuiltin*,
-        range_check_ptr,
-    }(
-        owners_index : felt,
-        owners_len : felt,
-        owners : felt*,
-    ):
+func _get_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_index : felt, owners_len : felt, owners : felt*
+):
     if owners_index == owners_len:
         return ()
     end
 
-     # Write the current iteration to storage
+    let (owner) = _owners.read(index=owners_index)
+    assert owners[owners_index] = owner
+
+    _get_owners(owners_index=owners_index + 1, owners_len=owners_len, owners=owners)
+    return ()
+end
+
+func _set_owners_impl{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*
+):
+    alloc_locals
+    require_unique_owners(owners_len, owners)
+
+    # Clean previous owners
+    let (old_owners_len) = _owners_len.read()
+    _clean_owners(0, old_owners_len)
+
+    let (local confirmations_required) = _confirmations_required.read()
+    let (lt) = is_le(owners_len, confirmations_required - 1)
+    if lt == TRUE:
+        # Lower number of confirmation
+        _confirmations_required.write(owners_len)
```

Comment:
Shouldn't this, instead, revert? If someone tries to set 2 owners and 5 confirmations, the code simply silently changes 5 to 2. I would prefer the code not to do unexpected things silently. Setting 2 owners and 5 confirmation should be an error.

Reply:
Making this part of the review.

Reply:
Ah you are correct, I didn't follow the logic about `set_owners_and_confirmations_required`. It will indeed revert if trying to set incompatible combination.
I don't have a very strong opinion about what should happen if amount of owner is lowered below required confirmation amount - should it revert or adjust the confirmation amount automatically. Can you share any internal reasoning you have behind this, or is it just based on coralxyz? In general I prefer to not have this kind of silent adjustments by the code and would maybe prefer a revert.

P.S. I'm very open to discussing these in a call or chat as well, if that makes things easier. Feel free to poke at Telegram @Lauri_P

Reply:
Yeah, it's a simple workflow that was inspired by audited [Serum multisig](https://github.com/coral-xyz/multisig/blob/a413b76de3497273a0cd27bcba6c1ac439053b67/programs/multisig/src/lib.rs#L127) (now rebranded as Coral I guess).

The idea is:

- Owners and threshold can be set simultaneously (constructor, setter) or separately (setters).
- Owners can be added or removed. If removed so the total length is under the threshold, the threshold is updated to equal the new owners' length.
- Threshold can be increased or decreased, it should always be in the range `[1, owners_len]`

Reply:
as I said, no strong opinion here. I'm ok with this